### PR TITLE
Feature/selectable middleware

### PR DESCRIPTION
**Pull request by teemukataja**
### Background
We have use cases at CSC where we are using `sda-download` in three different contexts, and each of them have a different authentication middleware. We have been maintaining three separate implementations, and would now like to streamline them into a single codebase. These changes here would lay the foundation for the capabilities we need (changing middleware, more versatile cache), and then we would implement our needs in our private repository.

### Changes
- new config `app.middleware`, with a default value `default`
- `TokenMiddleware` is now the default middleware, custom middlewares can be built in the `middleware` package, and be selected for runtime using the config (added a readme to the middleware directory for instructions)
- session cache now works with the whole cache struct instead of only the datasets slice, so that it can be used to store more items
- no breaking changes, only refactoring to allow us to use the same codebase for more use cases! :smile:

### CODE REVIEW

**Review by blankdots — APPROVED** [review]

File: api/middleware/middleware.go [hunk 1]
```diff
@@ -92,10 +95,13 @@ func storeDatasets(c *gin.Context, datasets []string) *gin.Context {
 }
 
 // GetDatasets extracts the dataset list from the request context
-var GetDatasets = func(c *gin.Context) []string {
-	datasets := c.GetStringSlice(datasetsKey)
-
-	log.Debugf("returning %v from request context", datasets)
+var GetDatasets = func(c *gin.Context) session.DatasetCache {
```
Comment (pontus) [inline api/middleware/middleware.go:98]: I don't know why it was a `var` but I don't see any reason to have it as such now, change to common `func` ?

**Review by pontus — APPROVED** [review]
Looks OK.

Could possibly have a short explainer in middleware.go to clarify the cache flow (at least I got tripped up a few times thinking about it), but I'm fine with merging as is.

File: api/middleware/middleware.go (hunk 1 above)
Comment (teemukataja) [reply api/middleware/middleware.go:98]: the `var` allows the function to be easily mocked by replacing the internals in the tests like this https://github.com/neicnordic/sda-download/blob/feature/selectable-middleware/api/sda/sda_test.go#L27-L32

Comment (pontus) [reply api/middleware/middleware.go:98]: Ah, yes, makes sense.

**Comment by teemukataja** [comment]
> Looks OK.
> 
> Could possibly have a short explainer in middleware.go to clarify the cache flow (at least I got tripped up a few times thinking about it), but I'm fine with merging as is.

I agree, I'll see if I can refactor it a little bit after this PR to simplify it

**Pull request merged**